cpp/example_code/s3-crt/s3-crt-demo.cpp

```diff
@@ -0,0 +1,227 @@
+#include <aws/s3-crt/S3CrtClient.h>
+#include <aws/s3-crt/model/CreateBucketRequest.h>
+#include <aws/s3-crt/model/DeleteBucketRequest.h>
+#include <aws/s3-crt/model/PutObjectRequest.h>
+#include <aws/s3-crt/model/GetObjectRequest.h>
+#include <aws/s3-crt/model/DeleteObjectRequest.h>
+
+static const char ALLOCATION_TAG[] = "s3-crt-demo";
+
+// List all Amazon S3 buckets under the account.
+bool ListBuckets(const Aws::S3Crt::S3CrtClient& s3CrtClient, const Aws::String& bucketName) {
+
+    Aws::S3Crt::Model::ListBucketsOutcome outcome = s3CrtClient.ListBuckets();
+
+    if (outcome.IsSuccess()) {
+        std::cout << "All buckets under my account:" << std::endl;
+
+        for (auto const& bucket : outcome.GetResult().GetBuckets())
+        {
+            std::cout << "  * " << bucket.GetName() << std::endl;
+        }
+        std::cout << std::endl;
+
+        return true;
+    }
+    else {
+        std::cout << "ListBuckets error:\n"<< outcome.GetError() << std::endl << std::endl;
+
+        return false;
+    }
+}
+
+// Create an Amazon S3 bucket.
+bool CreateBucket(const Aws::S3Crt::S3CrtClient& s3CrtClient, const Aws::String& bucketName) {
+
+    std::cout << "Creating bucket: \"" << bucketName << "\" ..." << std::endl;
+
+    Aws::S3Crt::Model::CreateBucketRequest request;
+    request.SetBucket(bucketName);
+
+    Aws::S3Crt::Model::CreateBucketOutcome outcome = s3CrtClient.CreateBucket(request);
+
+    if (outcome.IsSuccess()) {
+        std::cout << "Bucket created." << std::endl << std::endl;
+
+        return true;
+    }
+    else {
+        std::cout << "CreateBucket error:\n" << outcome.GetError() << std::endl << std::endl;
+
+        return false;
+    }
+}
+
+// Delete an existing Amazon S3 bucket.
+bool DeleteBucket(const Aws::S3Crt::S3CrtClient& s3CrtClient, const Aws::String& bucketName) {
+
+    std::cout << "Deleting bucket: \"" << bucketName << "\" ..." << std::endl;
+
+    Aws::S3Crt::Model::DeleteBucketRequest request;
+    request.SetBucket(bucketName);
+
+    Aws::S3Crt::Model::DeleteBucketOutcome outcome = s3CrtClient.DeleteBucket(request);
+
+    if (outcome.IsSuccess()) {
+        std::cout << "Bucket deleted." << std::endl << std::endl;
+
+        return true;
+    }
+    else {
+        std::cout << "DeleteBucket error:\n" << outcome.GetError() << std::endl << std::endl;
+
+        return false;
+    }
+}
+
+// Put an Amazon S3 object to the bucket.
+bool PutObject(const Aws::S3Crt::S3CrtClient& s3CrtClient, const Aws::String& bucketName, const Aws::String& objectKey, const Aws::String& fileName) {
+
+    std::cout << "Putting object: \"" << objectKey << "\" to bucket: \"" << bucketName << "\" ..." << std::endl;
+
+    Aws::S3Crt::Model::PutObjectRequest request;
+    request.SetBucket(bucketName);
+    request.SetKey(objectKey);
+    std::shared_ptr<Aws::IOStream> bodyStream = Aws::MakeShared<Aws::FStream>(ALLOCATION_TAG, fileName.c_str(), std::ios_base::in | std::ios_base::binary);
```

Comment:
1. SetBody() requires a shared pointer.
2. It's a shared pointer because we pass this pointer down to the http client. For async call, it will be shared to that thread.